### ISSUE TITLE
Not enough information on broken git repos

### ISSUE THREAD

**Issue by traxys** [nabijaczleweli/cargo-update#139]
When I run `cargo install-update --git --all` I get
```
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Error { code: -1, klass: 4, message: "corrupted loose reference file: FETCH_HEAD" }', /home/traxys/.cargo/registry/src/github.com-1ecc6299db9ec823/cargo-update-4.0.0/src/ops/mod.rs:449:13
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

This does not indicate to me at all where is the problem so that I could try to fix it, couldn't there be a bit more diagnostic there ?

**Comment by nabijaczleweli** [comment]
I didn't really expect that to be able to error; can you try the current `master` branch with af34089e706d311ba7c897283a60586aab4098ca?

**Comment by traxys** [comment]
It gives me the repo where there is a problem
```
Updated 1 package.
thread 'main' panicked at 'Updating HEAD in /tmp/cargo-update/gitoxide from https://github.com/Byron/git-oxide: corrupted loose reference file: FETCH_HEAD; class=Reference (4)', src/ops/mod.rs:451:30
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

**Comment by nabijaczleweli** [comment]
Ah, that's because this assumes that the default branch is always `master`. Would you mind uploading the result of `tar -caf gitoxide.tgz /tmp/cargo-update/gitoxide` here so I can poke at it and correctly read the default branch instead?

**Comment by traxys** [comment]
Uh I should have done that ..... `rm -rf`'d the thing :/ I need to learn to not delete everything so fast (never had to debug much --')

**Comment by nabijaczleweli** [comment]
It's quite alright, getting git repositories to the right state isn't *that* much of a pain in the arse.

But the error didn't persist after purging `/tmp/cargo-update`? That's good to know, means branchless fresh clones are working at least.

**Comment by nabijaczleweli** [comment]
@traxys can you build with 66655eaac8d5163cb1ec86de8809c2b0c2311d7d and run `cargo install-update -lag` a few times? I fixed the default branch handling (and, well, actually correctly update the repos instead of setting a HEAD), so it should work now.

**Comment by traxys** [comment]
So firstly, with af34089 I does not work after the initial clone. Now it panics every time on `thread 'main' panicked at 'HEAD not symbolic?', src/ops/mod.rs:443:21`

**Comment by nabijaczleweli** [comment]
Does purging `/tmp/cargo-update` help? Pre-https://github.com/nabijaczleweli/cargo-update/commit/66655eaac8d5163cb1ec86de8809c2b0c2311d7d updated repositories were left in a (moderately) broken state; purging and re-updating should fix this.

**Comment by traxys** [comment]
`rm -rf /tmp/cargo-update/*; cargo install-update -lag` gives the error :/

**Comment by nabijaczleweli** [comment]
Ugh, I can't reproduce this. 59be1c0de283dabce320a860a3d533d00910a6a9 should print the location in that error as well; maybe it's not under `/tmp/cargo-update`?

**Comment by traxys** [comment]
I have a bit more context on it
```
cargo install-update --git cargo-update
Package  Installed  Latest  Needs update

No packages need updating.
thread 'main' panicked at 'HEAD not symbolic?', src/ops/mod.rs:443:21
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace```
fails but 
```
❯ cargo install-update --git gitoxide
    Updating registry 'https://github.com/rust-lang/crates.io-index'

Package   Installed  Latest  Needs update
gitoxide  v0.1.0     v0.1.0  No

No packages need updating.
Package   Installed                                 Latest                                    Needs update
gitoxide  893f65b63b424922b8cdc496a9e798acc498c1c6  893f65b63b424922b8cdc496a9e798acc498c1c6  No

No git packages need updating.
Overall updated 0 packages.
```

**Comment by traxys** [comment]
I'll test with the latest branch when I have compiled it, I got that error when having trying to update cargo-update

**Comment by traxys** [comment]
```
thread 'main' panicked at 'HEAD not symbolic in /home/traxys/.cargo/git/db/cargo-depgraph-f9aefec71dabe73d', src/ops/mod.rs:447:44
```
with master

**Comment by nabijaczleweli** [comment]
Ah yes, that'll do it. This means we need to handle the upgrade gracefully instead of just giving up with a migration note. I'll have something for you tomorrow.

**Comment by nabijaczleweli** [comment]
Current `master` (at least 5958d07c21b7e38b6d11828b26c6aad16c5a01cd) should fix the problem; mind trying it?

**Comment by traxys** [comment]
Yes that fixes it ! Tough it hangs for a long time (some repo may take some time to clone), someone could mistake this for a problem

**Comment by nabijaczleweli** [comment]
Subsequent runs should be fine, though, right?

**Comment by traxys** [comment]
Yes I did a `-ag` run and it worked just fine

**Comment by nabijaczleweli** [comment]
dope

**Comment by nabijaczleweli** [comment]
Released in [`v4.1.0`](https://github.com/nabijaczleweli/cargo-update/releases/tag/v4.1.0), thanks for your help with this!
